date: fix interaction of flags, fix issues around --set

BenWiederhake opened the pull request:
This PR fixes multiple issues around multi-flags:
- Many repeated flags used to be handled incorrectly; this is work toward #5998 
- We used to accept the `--reference` argument, but completely ignored it. Yikes!
- Input, parsing, and output of `--set` was completely broken. This fixes it largely.
- Input options (`-d`, `-f`, `-r`), formatting options (`-I`, `-R`, `--rfc-3339`), and `--set` are **often** mutually-exclusive, with the exception that input options override themselves (but not each other), and also `--set` can be combined with formatting options.

Some things are unfixable, in particular the order-sensitive behavior; see [#4254#issuecomment-2026446634](https://github.com/uutils/coreutils/issues/4254#issuecomment-2026446634) and https://github.com/uutils/uutils-args/pull/113

BenWiederhake commented:
"CICD / Build (macos-latest, x86_64-apple-darwin, feat_os_macos)" flaked:
```
---- test_stat::test_stdin_pipe_fifo1 stdout ----
run: /Users/runner/work/coreutils/coreutils/target/x86_64-apple-darwin/debug/coreutils stat -
run: /Users/runner/work/coreutils/coreutils/target/x86_64-apple-darwin/debug/coreutils stat -L -
thread 'test_stat::test_stdin_pipe_fifo1' panicked at ''  File: -
  Size: 0         	Blocks: 0          IO Block: 512    weird file
Device: 25d29ea1h/634560161d	Inode: 2178033826  Links: 0
Access: (0660/?rw-rw----)  Uid: (  501/  runner)   Gid: (   20/   staff)
Access: 2024-03-29 16:28:42.942290340 +0000
Modify: 2024-03-29 16:28:42.942290340 +0000
Change: 2024-03-29 16:28:42.942290340 +0000
 Birth: 1970-01-01 00:00:00.000000000 +0000
' does not contain 'fifo'', tests/by-util/test_stat.rs:286:10
```

All other failures appear to be known issues, too.

BenWiederhake commented:
Changes since last push:
- Rebased on current main
- Resolved trivial conflict in `test_date.rs` with #4657

BenWiederhake commented:
Changes since last push:
- While resolving the conflict, I accidentally dropped the `#[test]` marker for a test. Whoops! Added it back in.

BenWiederhake commented:
Changes since last push:
- Rebased to new main, due to conflict with #6160

BenWiederhake commented:
Changes since last push: Nothing, just a rebase.

BenWiederhake commented:
Changes since last push:
- Rebased to new main
- Nothing more, I just want to remind that this PR exists and is still a good idea.

What else needs to be done?

BenWiederhake commented:
Changes since last push:
- Rebase on current main (that's around 100 commits!)
- Split the gargantuan `uumain` function into four separate functions, as requested, and remove the corresponding clippy lint.

The only CI failure is #6534, as usual.

sylvestre commented:
@BenWiederhake sorry for the latency, it needs a small rebase

sylvestre commented:
@BenWiederhake ping ? :)

BenWiederhake commented:
Changes since last summary:
- Rebase on current main, make sure to keep all the changes to `main` since then, even in case of refactored code.
- Appease clippy

The only CI failure is https://github.com/uutils/coreutils/issues/6534, as usual.

cakebaker commented on tests/by-util/test_date.rs:
```diff
@@ -259,6 +285,42 @@ fn test_date_set_permissions_error() {
+            // TODO "02/03/2001",
+        ] {
+            let result = new_ucmd!().arg("-s").arg(date_string).fails();
+            // stdout depends on the specific date; don't check it.
```
Here it might make sense to test that stdout is not empty because currently there is a bug in our implementation: we don't output anything to stdout whereas GNU `date` outputs the specified date.
```
$ cargo run -q date --set "yesterday"
date: cannot set date: Permission denied
$ date --set "yesterday"
date: cannot set date: Operation not permitted
Tue Feb 18 05:06:33 PM CET 2025
```
```suggestion
            // stdout depends on the specific date; don't check it.
            assert!(!result.stdout_str().is_empty());
```

BenWiederhake commented:
Changes since last summary:
- Rebase on current main
- Add stdout test suggested by @cakebaker – thanks!

This reveals two bugs in the PR:
- If `--set` doesn't have sufficient permissions, the error is raised first (and aborts the program), so no output occurs. Doink!
- If `--set` *does* have sufficient permissions, the timezone is wrong.

sylvestre commented:
jobs are failing (and needs some small rebases)
